### PR TITLE
feat(devtools): implement inspect functionality for directives

### PR DESCRIPTION
**Pull request by sbrshkappa**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
Devtools Feature

## What is the current behavior?
Previously, you could inspect the source code of a component but not a directive.

Issue Number: #45882

## What is the new behavior?
 This commit adds functionality to inspect source code for directives as well. Now you will see the inspect icon on the header component of each directive on a selected element.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-explorer.component.ts [hunk 1]
```diff
@@ -139,11 +139,20 @@ export class DirectiveExplorerComponent implements OnInit, OnDestroy {
...
     }
-    this._appOperations.viewSource(this.currentSelectedElement.position);
+
+    const directiveIndex = this.currentSelectedElement.directives.findIndex(
+        directive => directive.name === directiveName);
+
+    if (directiveIndex < 0)
+      this._appOperations.viewSource(this.currentSelectedElement.position);
+    else
+      this._appOperations.viewSource(this.currentSelectedElement.position, directiveIndex);
```
Comment (AleksanderBodurri) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-explorer.component.ts:155]: Small nit

```suggestion
    if (directiveIndex ===  -1) {
      // view the component definition
      this._appOperations.viewSource(this.currentSelectedElement.position);
      return;
   }
    
    // view the directive definition
   this._appOperations.viewSource(this.currentSelectedElement.position, directiveIndex);
```

File: devtools/projects/shell-browser/src/app/chrome-window-extensions.ts [hunk 2]
```diff
@@ -25,12 +25,21 @@ const extendWindowOperations = <T extends {}>(target, classImpl: T) => {
...
     const node = findNodeFromSerializedPosition(serializedId);
     if (node === null) {
       console.error(`Cannot find element associated with node ${serializedId}`);
       return;
     }
+    if (directiveIndex !== undefined) {
+      if (node.directives[directiveIndex]) {
+        return node.directives[directiveIndex].instance.constructor;
+      } else {
+        console.error(
```
Comment (AleksanderBodurri) [inline devtools/projects/shell-browser/src/app/chrome-window-extensions.ts:39]: Lets add an early return after this line so we never run the code underneath it

**Review by AleksanderBodurri — CHANGES_REQUESTED** [review]
It was nice meeting you at NG Conf 😄 Thanks for the PR. I left a couple of comments to be resolved but overall this looks good to me

**Comment by sbrshkappa** [comment]
> It was nice meeting you at NG Conf 😄 Thanks for the PR. I left a couple of comments to be resolved but overall this looks good to me

Absolutely! Great meeting you as well. Helps to work in person, to get things moving quickly.

File: devtools/projects/shell-browser/src/app/chrome-window-extensions.ts (hunk 2 above)
Comment (sbrshkappa) [reply devtools/projects/shell-browser/src/app/chrome-window-extensions.ts:39]: Added a return here. Is that what you meant?

**Comment by sbrshkappa** [comment]
@AleksanderBodurri I have the changes merged, I did a combination of using the github "Apply changes" and committing the other changes from my local. Hopefully the branch looks good. Let me know.

**Comment by sbrshkappa** [comment]
@AleksanderBodurri , I accidentally did some merge that messed up this branch, can you help with fixing this?

**Comment by AleksanderBodurri** [comment]
Backup your branch just in case
```bash
git checkout -b view-source-directive-bkp
git checkout sk-view-source-directives
```

You can reset to a branch with only your commits by:
```bash
# reset softly to the commit right before 7f3180e and f86b36d
git reset --soft 0dc66120f053e610398cbdba61b6951ce77ee5b4
# stash the changes from 7f3180e and f86b36d
git stash
# reset hard to right before the merge 
git reset --hard 3cf56797ca
# pop stash to bring back the changes from 7f3180e and f86b36d
git stash pop
# make new commit with the changes from 7f3180e and f86b36d
git add .
git commit
# update remote branch
git push -f
```

**Comment by sbrshkappa** [comment]
Thank you for the detail steps, @AleksanderBodurri. The branch should be back in the original state.

**Comment by AleksanderBodurri** [comment]
Awesome 😄 The last thing I would suggest is that you squash your 3 commits into 1. You can do this anyway you're comfortable with, I'll leave a description of how I would do this below for reference.

Run the following in your terminal
```
git rebase -i main
```
and then change the `pick` next to [0597b8f](https://github.com/angular/angular/pull/47334/commits/0597b8f3daedf283f5562098322e7a955915f567) and [3cf5679](https://github.com/angular/angular/pull/47334/commits/3cf56797ca0ff1f71c88c98e5367ac642b912aa6) to `fixup` or `f`. Saving this file will cause the rebase to fixup these 2 commits onto [b161d6b](https://github.com/angular/angular/pull/47334/commits/b161d6bed527f79a4ff98abc9e6c953fe369b92f). Then update your remote branch with `git push -f`.

**Comment by sbrshkappa** [comment]
@AleksanderBodurri . Done!

**Review by AleksanderBodurri — APPROVED** [review]
LGTM

**Review by AleksanderBodurri — APPROVED** [review]

**Review by AleksanderBodurri — APPROVED** [review]

**Review by AleksanderBodurri — APPROVED** [review]
Reviewed-for: devtools

**Review by josephperrott — APPROVED** [review]
LGTM

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit 32cad55f0d6ad2af6841e05e26a97bad9abc17cc.

**Pull request closed without merging**